File: 백준/Silver/11055. 가장 큰 증가하는 부분 수열/가장 큰 증가하는 부분 수열.py
n = int(input())
li = list(map(int, input().split()))
x = [li[0]] * len(li) #부분계산값 저장할 곳
for i in range(1,len(li)):
    m = 0
    for j in range(0,i):
        if(li[j]<li[i]):
            m = max(m, x[j])
    x[i] = m + li[i]
print(max(x))

File: 백준/Silver/11055. 가장 큰 증가하는 부분 수열/가장 큰 증가하는 부분 수열.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = Integer.parseInt(sc.nextLine());
        String[] A_ = sc.nextLine().split(" ");
        int[] A = new int[N];
        for(int i=0; i<N; i++){
            A[i] = Integer.parseInt(A_[i]);
        }
        int[] dp = new int[N];
        for (int i = 0; i < N; i++) {
            dp[i] = A[i];  
        }
        for(int i=0; i<N; i++){
                for(int j=i;j>=0;j--){
                    if(A[j]<A[i]){
                        dp[i] = Math.max(dp[i], dp[j]+A[i]);
                    }
                }
            }
              
        int answer = 0;
        for(int d:dp){
            answer = Math.max(answer,d);
        }
        
        System.out.println(answer);
        // System.out.println(Arrays.toString(dp));
        }
}
